File: ts/fuzzySearch.ts
import { genArr } from "./array";

export type FuzzySearchMatch = {
    index: number,
    length: number,
    score: number,
    minimumEditDistance: number,
}

let seedRow = genArr(8, i => i);
let currentRow = genArr(8, 0);
let nextCurrentRow = genArr(8, 0);

export function fuzzySearch<
    MS extends (number | undefined) = undefined,
    MED extends (number | undefined) = undefined
>(
    test: string,
    search: string,
    other?: {
        minimumScore?: MS,
        maximumEditDistance?: MED,
        startingIndex?: number,
        mode?: "bestMatch" | "firstMatch"
        /**
         * @param index The index that is about to be searched.
         * @param bestMatch The best match found so far.
         * @returns If true: stops searching.
         */
        onProgress?: (index: number, bestMatch: FuzzySearchMatch) => boolean | undefined
    }
): (MS extends number ? (MS extends 0 ? never : undefined)
        : MED extends number ? undefined
            : never)
| FuzzySearchMatch {
    type Return = ReturnType<typeof fuzzySearch<MS, MED>>;
    //#region parameters
    /** minimum score allowed */
    const minimumScore = other?.minimumScore ?? 0;
    /** maximum edit distance allowed */
    const maximumEditDistance = other?.maximumEditDistance ?? Infinity;
    const startingIndex = other?.startingIndex ?? 0;

    /** minimum number of matching characters that achieves the minimum score */
    const minimumMatchesForScore = Math.ceil(search.length * minimumScore);
    /** maximum edit distance that achieves the minimum score */
    const maximumEditDistanceForScore = search.length - minimumMatchesForScore;

    /** maximum edit distance allowed that achieves the minimum score */
    const appliedMaximumEditDistance = Math.min(
        maximumEditDistance,
        maximumEditDistanceForScore,
    )
    //#endregion

    // find the first match
    //    ... the empty string
    /** index of best match so far */
    let index = 0;
    /** length of best match so far */
    let length = 0;
    /** min edit dist of best match so far */
    let minimumEditDistance = search.length;

    //#region easy guards
    if (search === "" || test === "") {
        return {
            index,
            length,
            score: search === "" ? 1 : 0,
            minimumEditDistance
        };
    }
    // /!\ at this point it's assumed search and test both contain at least 1 character /!\

    if (search.length - test.length > appliedMaximumEditDistance){
        return undefined as Return;
    }

    // indexOf is 2 orders of magnitude faster than this function
    // this means it can check for a perfect match ahead of time
    // and if it's a perfect match, then fuzzy searching won't find anything better
    // so short circuit and return here
    const indexOf_index = test.indexOf(search);
    if (indexOf_index >= 0) {
        return {
            index: indexOf_index,
            length: search.length,
            minimumEditDistance: 0,
            score: 1
        }
    } else if (appliedMaximumEditDistance === 0) {
        return undefined as Return;
    }

    // /!\ at this point it's known that the lowest edit distance to be found is 1 /!\
    //#endregion

    //     a b c d e f <- search string
    //  [0 1 2 3 4 5 6] <- seedRow (and initial prevRow)
    // a 1 0 1 2 3 4 5
    // z[2 1 1 2 3 4 5] <- prevRow
    // c[3 2 2 1 2 3 4] <- currentRow
    // f 4 3 3 2 2 3(3) <- minimum edit distance for length of substring so far
    // g 5 4 4 3 3 3(4) <- minimum edit distance for length of substring so far
    // g 6 5 5 4 4 4(4) <- minimum edit distance for length of substring so far
    // ^
    //  ` - substring of test
    const columnCount = search.length + 1;
    if (seedRow.length < columnCount) {
        seedRow = genArr(columnCount, i => i);
        currentRow = genArr(columnCount, 0);
        nextCurrentRow = genArr(columnCount, 0);
    }
    let prevRow = seedRow

    /** the best possible edit distance to find at the window's current position */
    let potentialEditDistanceForI = 0;
    substringIndex: for (let i = startingIndex; i < test.length; i++) {
        // check if this window position is worth searching
        if (
            potentialEditDistanceForI >= minimumEditDistance
            || potentialEditDistanceForI > appliedMaximumEditDistance
        ) {
            potentialEditDistanceForI -= 2;
            continue;
        }

        //#region update progress
        const stop = other?.onProgress?.(
            i,
            {index, length, minimumEditDistance, score: calcScore(minimumEditDistance)}
        );
        if (stop) break;
        //#endregion

        /** the best edit distance found at this window position
         * used to update `potentialEditDistanceForI`
         */
        let minimumEditDistanceFromI = Infinity;
        let lOfMinimumEditDistanceFromI = 0;

        //#region expand window
        /** length of substring of test so far */
        let l = 1;
        for (; l <= test.length - i; l++) {
            /** index within test */
            const ti = i + l - 1;
            /** the current row in the matrix */
            const r = l;


            //     a b c d e f
            //   0 1 2 3 4 5 6
            // a 1 0 1 2 3 4 5
            // z 2 1 1 2 3 4 5
            // c 3 2 2 1 2 3 4
            // f 4 3 3 2 2 3 3
            //   ^
            //    `- populate the seed column
            currentRow[0] = r;

            /** length of search string so far (the column of the matrix) */
            let sl = 1;
            for (;sl <= search.length; sl++) {
                /** index within search */
                let si = sl - 1;
                /** the current column in the matrix */
                const c = sl;

                const searchChar = search.charAt(si);
                const testChar = test.charAt(ti);
                if (searchChar === testChar) {
                    currentRow[c] = prevRow[c - 1];
                } else {
                    const north = prevRow[c];
                    const northWest = prevRow[c - 1];
                    const west = currentRow[c - 1];
                    currentRow[c] = 1 + Math.min(north, northWest, west);
                }
            }

            const editDist = currentRow[search.length];

            prevRow = currentRow;


            // rotate rows
            let temp = currentRow;
            currentRow = nextCurrentRow;
            nextCurrentRow = temp;
            // const editDist = currentRow[search.length];
            if (editDist <= minimumEditDistanceFromI) {
                minimumEditDistanceFromI = editDist
                lOfMinimumEditDistanceFromI = l
            }

            // clamp window size
            /** The best potential edit distance to be found by continuing this loop */
            const potentialEditDist = editDist - (search.length - l);
            if (
                potentialEditDist > minimumEditDistance
                || potentialEditDist > appliedMaximumEditDistance
            ){
                break;
            }
        }
        //#endregion
        if (minimumEditDistanceFromI < minimumEditDistance) {
            // better match found

            minimumEditDistance = minimumEditDistanceFromI;
            index = i;
            length = lOfMinimumEditDistanceFromI;
        }

        // edit distance can't get better than 1 thanks to the indexOf check at the start
        if (minimumEditDistance <= 1) break substringIndex;

        //#region firstMatch mode
        if (other?.mode === "firstMatch" && minimumEditDistance <= appliedMaximumEditDistance) {
            const score = calcScore(minimumEditDistance)
            if (score >= minimumScore) {
                if (length > search.length){
                    const bestWithinFirst = fuzzySearch(
                        test.substring(
                            index,
                            // TODO is this right?
                            index + length + minimumEditDistance
                        ),
                        search,
                        {
                            maximumEditDistance: maximumEditDistance,
                            minimumScore: minimumScore,
                        }
                    );
                    if (bestWithinFirst === undefined) {
                        return {index, length, minimumEditDistance, score}
                    } else {
                        return {
                            ...bestWithinFirst,
                            index: index + bestWithinFirst?.index
                        }
                    }
                } else {
                    return {index, length, minimumEditDistance, score}
                }
            }
        }
        //#endregion

        // reset rows
        prevRow = seedRow;

        // potential edit distance of next window position
        if (minimumEditDistanceFromI === Infinity) {
            potentialEditDistanceForI -= 2;
        } else {
            potentialEditDistanceForI = minimumEditDistanceFromI - 2;
        }
    }

    const score = calcScore(minimumEditDistance);

    if (score >= (minimumScore) && minimumEditDistance <= (maximumEditDistance)) {
        return {index, length, score, minimumEditDistance};
    } {
        return undefined as Return;
    }


    function calcScore(editDistance: number) {
        // it is assumed that search.length is > 0
        const matchCount = search.length - editDistance
        const result = matchCount / search.length;
        return result;
    }
}

File: go/fuzzySearch.go
package main

import (
	"math"
	"sync"
	"unicode/utf8"
)

type FuzzySearchMatch struct {
	minimumEditDistance int
	score               float64
	runeOffset          int
	byteOffset          int
	runeCount           int
	byteCount           int
}

type FuzzySearchParams struct {
	testString          string
	searchString        string
	minimumScore        float64
	maximumEditDistance int
	cacheDepth          int
	takeProgress        func(FuzzySearchProgress) bool
	rootCache           *rowCache
}

type FuzzySearchProgress struct {
	byteOffset int
	runeOffset int
	bestMatch  FuzzySearchMatch
}

func FuzzySearch(test, search string) FuzzySearchMatch {
	return FuzzySearchWith(test, search).Run()
}
func FuzzySearchWith(test, search string) FuzzySearchParams {
	return FuzzySearchParams{
		testString:          test,
		searchString:        search,
		minimumScore:        0.0,
		maximumEditDistance: math.MaxInt,
		cacheDepth:          2,
	}
}

func (self FuzzySearchParams) Run() FuzzySearchMatch {
	result := fuzzySearchFromBuilder(self)
	return result
}

func (self FuzzySearchParams) MinScore(value float64) FuzzySearchParams {
	self.minimumScore = value
	return self
}

func (self FuzzySearchParams) MaxEditDistance(value int) FuzzySearchParams {
	self.maximumEditDistance = value
	return self
}

func (self FuzzySearchParams) CacheDepth(value int) FuzzySearchParams {
	self.cacheDepth = value
	return self
}

func (self FuzzySearchParams) TakeProgress(value func(FuzzySearchProgress) bool) FuzzySearchParams {
	self.takeProgress = value
	return self
}

type rowCache struct {
	row      []int
	children sync.Map
}

var nilCache *rowCache = &rowCache{}

func calcScore(editDistance, searchLength int) float64 {
	matchCount := searchLength - editDistance
	score := float64(matchCount) / float64(searchLength)
	return score
}

func fuzzySearchFromBuilder(params FuzzySearchParams) FuzzySearchMatch {
	//#region take options
	test := params.testString
	search := params.searchString
	minimumScore := 0.0
	maximumEditDistance := math.MaxInt
	cacheDepth := params.cacheDepth
	minimumScore = max(0.0, min(1.0, params.minimumScore))
	maximumEditDistance = max(0, params.maximumEditDistance)
	takeProgress := params.takeProgress
	rootCache := params.rootCache
	if rootCache == nil {
		rootCache = &rowCache{}
	}
	//#endregion

	searchLength := utf8.RuneCountInString(search)

	minimumMatchesForMinimumScore := math.Ceil(minimumScore * float64(searchLength))
	maximumEditDistanceForMinimumScore := searchLength - int(minimumMatchesForMinimumScore)

	appliedMaximumEditDistance := min(maximumEditDistance, maximumEditDistanceForMinimumScore)

	// default values for empty string
	minimumEditDistance := searchLength
	runeOffset := 0
	byteOffset := 0
	runeCount := 0
	byteCount := 0

	if search == "" || test == "" {
		var score float64
		if search == "" {
			score = 1.0
		} else {
			score = 0.0
		}

		return FuzzySearchMatch{
			minimumEditDistance: minimumEditDistance,
			score:               score,
			runeOffset:          runeOffset,
			byteOffset:          byteOffset,
			runeCount:           runeCount,
			byteCount:           byteCount,
		}
	}

	columnCount := searchLength + 1
	seedRow := make([]int, columnCount)
	row := make([]int, columnCount)
	nextRow := make([]int, columnCount)
	for i := range seedRow {
		seedRow[i] = i
	}

	prevRow := seedRow

	// window byte offset
	wb := 0
	// window rune offset
	wi := 0

	potentialEditDistanceForWI := 0
	for {
		// byte count of rune at start of window
		windowRuneLength, _ := bytesInRune_Utf8(test[wb])

		// byte count of current rune in test
		testRuneLength := windowRuneLength

		// edit distance of best substring in window (the one with the lowest ED)
		minimumEditDistanceFromWI := math.MaxInt
		// rune count of best substring in window
		lOfMinimumEditDistanceFromWI := 0
		// byte count of best substring in window
		bcOfMinimumEditDistanceFromWI := 0

		// window byte count
		wbc := 0

		// test byte offset
		tb := wb
		// test rune offset
		ti := wi
		// window length
		wl := 1

		prevCache := rootCache

		// check if it's worth checking the windows position
		if potentialEditDistanceForWI >= minimumEditDistance || potentialEditDistanceForWI > appliedMaximumEditDistance {
			// if not: skip to the next one
			goto nextWindowPosition // this is all the way down here because goto's can't jump over variable declarations
		}

		for {
			wbc += testRuneLength
			if wbc >= len(test)-wb {
				break
			}
			testRune, _ := runeAtByteInString(test, tb)
			// ED matrix row
			r := wl
			cacheAny, cacheHit := prevCache.children.Load(testRune)

			if cacheHit {
				cache := cacheAny.(*rowCache)
				prevRow = cache.row
				prevCache = cache
			} else {
				// populate seed column
				row[0] = r

				// search rune offset
				si := 0
				// search byte offset
				sb := 0
				for {
					searchRuneLength, _ := bytesInRune_Utf8(search[sb])
					c := si + 1

					// instead of converting the bytes in search and test
					// to runes, just check if the bytes match
					// #region check bytes for match
					match := searchRuneLength == testRuneLength

					if match {
						for rb := 0; rb < testRuneLength; rb++ {
							// searchByte := unsafeBoundlessStringGet(search, uintptr(sb+rb))
							// testByte := unsafeBoundlessStringGet(test, uintptr(tb+rb))
							// match = searchByte == testByte
							match = search[sb+rb] == test[tb+rb]
							if !match {
								break
							}
						}
					}

					// searchRune, _ := runeAtByteInString(search, sb)
					// match := testRune == searchRune
					//#endregion

					if match {
						// row[c] = prevRow[c-1]
						editDist := unsafeBoundlessSliceGet_int(prevRow, uintptr(c-1))
						unsafeBoundlessSliceSet_int(row, uintptr(c), editDist)
					} else {
						// north := prevRow[c]
						// northWest := prevRow[c-1]
						// west := row[c-1]

						// row[c] = 1 + min(north, northWest, west)

						north := unsafeBoundlessSliceGet_int(prevRow, uintptr(c))
						northWest := unsafeBoundlessSliceGet_int(prevRow, uintptr(c-1))
						west := unsafeBoundlessSliceGet_int(row, uintptr(c-1))

						unsafeBoundlessSliceSet_int(row, uintptr(c), 1+min(north, northWest, west))
					}

					si++
					sb += searchRuneLength
					if sb >= len(search) || si >= searchLength {
						break
					}
				}

				// populate cache
				if r <= cacheDepth {
					cacheRow := make([]int, len(row))
					copy(cacheRow, row)
					cache := &rowCache{row: cacheRow}
					// prevCache.children[testRune] = cache
					prevCache.children.Store(testRune, cache)
					prevCache = cache
				} else {
					prevCache = nilCache
				}
				// rotate rows
				prevRow = row
				temp := row
				row = nextRow
				nextRow = temp
			}

			// editDist := prevRow[searchLength]
			editDist := unsafeBoundlessSliceGet_int(prevRow, uintptr(searchLength))

			if editDist <= minimumEditDistanceFromWI {
				minimumEditDistanceFromWI = editDist
				lOfMinimumEditDistanceFromWI = wl
				bcOfMinimumEditDistanceFromWI = wbc
			}

			// clamp window size
			potentialEditDist := editDist - (searchLength - wl)
			if potentialEditDist > minimumEditDistance || potentialEditDist > appliedMaximumEditDistance {
				break
			}

			tb += testRuneLength
			ti++
			wl++

			testRuneLength, _ = bytesInRune_Utf8(test[tb])
		}
		// reset rows
		prevRow = seedRow

		if minimumEditDistanceFromWI < minimumEditDistance {
			minimumEditDistance = minimumEditDistanceFromWI
			byteOffset = wb
			runeOffset = wi
			byteCount = bcOfMinimumEditDistanceFromWI
			runeCount = lOfMinimumEditDistanceFromWI
		}

		// check for perfect match
		if minimumEditDistance == 0 {
			// perfect match found
			break
		}

		//#region progress report
		if takeProgress != nil {
			bestMatch := FuzzySearchMatch{minimumEditDistance: searchLength}

			if minimumEditDistance <= maximumEditDistance {
				score := calcScore(minimumEditDistance, searchLength)
				if score >= minimumScore {
					bestMatch.minimumEditDistance = minimumEditDistance
					bestMatch.score = score

					bestMatch.byteCount = byteCount
					bestMatch.byteOffset = byteOffset
					bestMatch.runeCount = runeCount
					bestMatch.runeOffset = runeOffset
				}
			}

			progress := FuzzySearchProgress{
				byteOffset: wb,
				runeOffset: wi,
				bestMatch:  bestMatch,
			}

			stop := takeProgress(progress)
			if stop {
				break
			}
		}
		//#endregion

	nextWindowPosition:
		// update potential edit distance for window position
		if minimumEditDistanceFromWI == math.MaxInt {
			potentialEditDistanceForWI -= 2
		} else {
			potentialEditDistanceForWI = minimumEditDistanceFromWI - 2
		}

		wb += windowRuneLength
		wi++

		if wb >= len(test) {
			break
		}
	}

	score := calcScore(minimumEditDistance, searchLength)

	if score < minimumScore || minimumEditDistance > appliedMaximumEditDistance {
		// remove undefined behavior
		result := FuzzySearchMatch{
			minimumEditDistance: searchLength,
		}

		return result
	}

	result := FuzzySearchMatch{
		minimumEditDistance: minimumEditDistance,
		score:               score,
		runeOffset:          runeOffset,
		byteOffset:          byteOffset,
		runeCount:           runeCount,
		byteCount:           byteCount,
	}

	return result
}
